bbq/vm/test/vm_test.go

```diff
@@ -7845,3 +7845,190 @@ func TestSaturatingArithmetic(t *testing.T) {
+
+		assert.Equal(
+			t,
+			interpreter.NewUnmeteredIntValueFromInt64(5),
+			result,
+		)
+	})
+
+	t.Run("update referenced var before first use", func(t *testing.T) {
+		t.Parallel()
+
+		result, err := CompileAndInvoke(t,
+			`
+              var a = 5
+              var b = a
+
+              fun test(): Int {
+                  // Update 'a' before getting 'b'.
+                  a = 8
+
+                  // Get 'b' for the fist time.
+                  return b
+              }
+            `,
+			"test",
+		)
+		require.NoError(t, err)
+
+		assert.Equal(
+			t,
+			interpreter.NewUnmeteredIntValueFromInt64(5),
+			result,
+		)
+	})
+
+	t.Run("update referenced var after first use", func(t *testing.T) {
+		t.Parallel()
+
+		result, err := CompileAndInvoke(t,
+			`
+              var a = 5
+              var b = a
+
+              fun test(): Int {
+                  // Use 'b' to get the value once.
+                  var c = b
+
+                  // Update 'a' before getting 'b' for the first time.
+                  a = 8
+
+                  // Get 'b' for a second time.
+                  // Should return the initial value.
+                  return b
+              }
+            `,
+			"test",
+		)
+		require.NoError(t, err)
+
+		assert.Equal(
+			t,
+			interpreter.NewUnmeteredIntValueFromInt64(5),
+			result,
+		)
+	})
+
+	t.Run("overridden local var", func(t *testing.T) {
+		t.Parallel()
+
+		result, err := CompileAndInvoke(t,
+			`
+              var a = 5
+
+              fun test(): Int {
+                  var a = 8
+                  return getGlobalA()
+              }
+
+              fun getGlobalA(): Int {
+                  return a
+              }
+            `,
+			"test",
+		)
+		require.NoError(t, err)
+
+		assert.Equal(
+			t,
+			interpreter.NewUnmeteredIntValueFromInt64(5),
+			result,
+		)
+	})
+
+	t.Run("forward references", func(t *testing.T) {
+		t.Parallel()
+
+		result, err, logs := CompileAndInvokeWithLogs(t,
+			`
+              var a = initializeA()
+              var b = initializeB()
+
+              fun test(): Int {
+                  log("invoked test")
+                  return a
+              }
+
+              fun initializeA(): Int {
+                  log("invoked initializeA")
+
+                  // Indirect forward reference to b
+                  var c = b
+
+                  log("exiting initializeA")
+                  return 5
+              }
+
+              fun initializeB(): Int {
+                  log("invoked initializeB")
+                  return 5
+              }
+            `,
+			"test",
+		)
+		require.NoError(t, err)
+
+		assert.Equal(
+			t,
+			interpreter.NewUnmeteredIntValueFromInt64(5),
+			result,
+		)
+
+		assert.Equal(
+			t,
+			[]string{
+				// Variables must be initialized before calling the function test.
+				`"invoked initializeA"`,
+				`"invoked initializeB"`,
+				`"exiting initializeA"`,
+
+				`"invoked test"`,
+			},
+			logs,
+		)
+	})
+}
```

Comment:
Could you please also add a test case which tests that the global variables are all force-evaluated at program start? Maybe just a variant of the last test where `test` is a no-op, and the global variable has a side-effect of logging.

Reply:
Added on 5885c1d381a38fe33ea0c591aebb5865998d04a8